### PR TITLE
Vt rewards distributor

### PR DESCRIPTION
**Pull request by verminator23**

### CODE REVIEW

File: contracts/cw-distributor/src/contract.rs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    let owner = deps.api.addr_validate(&msg.owner)?;
+    let recipient = deps.api.addr_validate(&msg.recipient)?;
+
+    let reward_token = deps.api.addr_validate(&msg.reward_token)?;
```
Comment (ezekiiel) [inline contracts/cw-distributor/src/contract.rs:28]: Might be nice to run some queries against these addresses when we set them to verify that these are indeed token / staking contracts. Could run `TotalStakedAtHeight` and `TokenInfo`.

File: contracts/cw-distributor/src/contract.rs [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    reward_token: String,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    if config.owner != info.sender {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let owner = deps.api.addr_validate(&owner)?;
+    let recipient = deps.api.addr_validate(&recipient)?;
+    let reward_token = deps.api.addr_validate(&reward_token)?;
```
Comment (ezekiiel) [inline contracts/cw-distributor/src/contract.rs:76]: Queries here as well if we decide we'd like to run them.

File: contracts/cw-distributor/src/contract.rs [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        owner,
+        recipient,
+        reward_token,
+        reward_rate,
+    };
+    CONFIG.save(deps.storage, &config)?;
+    Ok(Response::default())
+}
+
+pub fn execute_distribute(deps: DepsMut, env: Env) -> Result<Response, ContractError> {
```
Comment (ezekiiel) [inline contracts/cw-distributor/src/contract.rs:88]: This function will lock tokens in the contract is `config.reward_rate * block_diff` is greater than the contract's token balance. This is likely to happen when the contract is coming to the end of its reward period. How would you feel about making the amount to send the min of remaining balance and `amount`?

Comment (ezekiiel) [reply contracts/cw-distributor/src/contract.rs:88]: Likely want a test for this situation as well. :)

**Review by ezekiiel — COMMENTED** [review]
Looks great! Just one thing around locking the contract at the end of the rewards period and then I think this is good to go.

File: contracts/cw-distributor/src/contract.rs (hunk 3 above)
Comment (verminator23) [reply contracts/cw-distributor/src/contract.rs:88]: Great suggestion. I'll add that

File: contracts/cw-distributor/src/contract.rs (hunk 1 above)
Comment (verminator23) [reply contracts/cw-distributor/src/contract.rs:28]: @ben2x4 Lmk if you think this type of validation is necessary. I'm down to add it

File: contracts/cw-distributor/src/contract.rs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+use cosmwasm_std::{Binary, Deps, DepsMut, Env, MessageInfo, Response, StdResult};
+use cw2::set_contract_version;
+use stake_cw20::msg::ReceiveMsg;
+
+use crate::error::ContractError;
+use crate::msg::{ConfigResponse, ExecuteMsg, InstantiateMsg, LastPaymentBlockResponse, QueryMsg};
+use crate::state::{Config, CONFIG, LAST_PAYMENT_BLOCK};
+
+const CONTRACT_NAME: &str = "crates.io:cw-distributor";
```
Comment (ben2x4) [inline contracts/cw-distributor/src/contract.rs:13]: I would probable change name here and for file to `stake-cw20-reward-distributor` as I don't think the contract is very general purpose.

File: contracts/cw-distributor/src/contract.rs (hunk 3 above)
Comment (ben2x4) [reply contracts/cw-distributor/src/contract.rs:88]: yup, had same thought and +1 here

File: contracts/cw-distributor/src/contract.rs [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        contract_addr: config.reward_token.into(),
+        msg,
+        funds: vec![],
+    }
+    .into();
+    Ok(Response::default().add_message(send_msg))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
```
Comment (ben2x4) [inline contracts/cw-distributor/src/contract.rs:115]: I would consider combining all of this into one `info` response and including the contracts current balance of the reward token. Having reward token balance makes it easy for the caller to see if the rewards are active and calculate the remaining length without making a seperate query.

**Review by ben2x4 — COMMENTED** [review]
This is great, I think the minimal amount of state you have included here is really elegant. A couple of small suggestions and then ready to go!

File: contracts/cw-distributor/src/contract.rs (hunk 1 above)
Comment (ben2x4) [reply contracts/cw-distributor/src/contract.rs:28]: hmm, I know @ezekiiel  likes these types of checks, but i don't have a strong preference.

File: contracts/cw-distributor/src/contract.rs (hunk 4 above)
Comment (ben2x4) [reply contracts/cw-distributor/src/contract.rs:13]: this also make it clear it is part of the stake-cw20 family of contracts

File: contracts/cw-distributor/src/contract.rs (hunk 5 above)
Comment (verminator23) [reply contracts/cw-distributor/src/contract.rs]: Done

File: contracts/cw-distributor/src/contract.rs (hunk 4 above)
Comment (verminator23) [reply contracts/cw-distributor/src/contract.rs]: Done

File: contracts/cw-distributor/src/contract.rs (hunk 3 above)
Comment (verminator23) [reply contracts/cw-distributor/src/contract.rs]: Done

File: contracts/cw-distributor/src/contract.rs (hunk 1 above)
Comment (verminator23) [reply contracts/cw-distributor/src/contract.rs]: Added these checks + tests

**Comment by verminator23** [comment]
@ezekiiel @ben2x4 I've addressed all of your comments

**Review by ben2x4 — APPROVED** [review]
LGTM, amazing job here!

**Review by ezekiiel — APPROVED** [review]
:shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit:

**Comment by JakeHartnell** [comment]
LFG!

**Pull request merged**